File: python/nnet.py
import numpy as np

class NNet:
    """
    Class that represents a fully connected ReLU network from a .nnet file.
    """

    def __init__(self, filename):
        # Load network from the .nnet file
        with open(filename) as f:
            # Skip the comment lines starting with "//"
            line = f.readline()
            while line.startswith("//"):
                line = f.readline()

            # Read network metadata
            numLayers, inputSize, outputSize, _ = [int(x) for x in line.strip().split(",")[:-1]]
            layerSizes = [int(x) for x in f.readline().strip().split(",")[:-1]]
            symmetric = int(f.readline().strip().split(",")[0])

            # Read input normalization data
            inputMinimums = [float(x) for x in f.readline().strip().split(",")[:-1]]
            inputMaximums = [float(x) for x in f.readline().strip().split(",")[:-1]]
            inputMeans = [float(x) for x in f.readline().strip().split(",")[:-1]]
            inputRanges = [float(x) for x in f.readline().strip().split(",")[:-1]]

            # Read weights and biases for each layer
            weights = []
            biases = []
            for layernum in range(numLayers):
                previousLayerSize = layerSizes[layernum]
                currentLayerSize = layerSizes[layernum + 1]
                
                # Read weights
                weight_matrix = np.zeros((currentLayerSize, previousLayerSize))
                for i in range(currentLayerSize):
                    weight_matrix[i] = [float(x) for x in f.readline().strip().split(",")[:-1]]
                weights.append(weight_matrix)

                # Read biases
                bias_vector = np.zeros(currentLayerSize)
                for i in range(currentLayerSize):
                    bias_vector[i] = float(f.readline().strip().split(",")[0])
                biases.append(bias_vector)

            # Store network parameters
            self.numLayers = numLayers
            self.layerSizes = layerSizes
            self.inputSize = inputSize
            self.outputSize = outputSize
            self.mins = inputMinimums
            self.maxes = inputMaximums
            self.means = inputMeans
            self.ranges = inputRanges
            self.weights = weights
            self.biases = biases

    def evaluate_network(self, inputs):
        """
        Evaluate network using given inputs.
        """
        inputs = np.asarray(inputs).flatten()  # Ensure the input is flattened
        if len(inputs) != self.inputSize:
            raise ValueError(f"Expected input size {self.inputSize}, but got {len(inputs)}.")

        # Normalize inputs
        inputsNorm = np.array([
            (self.mins[i] if inputs[i] < self.mins[i] else self.maxes[i] if inputs[i] > self.maxes[i]
             else inputs[i] - self.means[i]) / self.ranges[i]
            for i in range(self.inputSize)
        ])

        # Forward pass through the network
        for layer in range(self.numLayers - 1):
            inputsNorm = np.maximum(np.dot(self.weights[layer], inputsNorm) + self.biases[layer], 0)
        outputs = np.dot(self.weights[-1], inputsNorm) + self.biases[-1]

        # Undo output normalization
        outputs = outputs * self.ranges[-1] + self.means[-1]
        return outputs

    def evaluate_network_multiple(self, inputs):
        """
        Evaluate network using multiple sets of inputs.
        """
        inputs = np.asarray(inputs)
        if inputs.ndim != 2 or inputs.shape[1] != self.inputSize:
            raise ValueError(f"Expected input shape (N, {self.inputSize}), but got {inputs.shape}.")

        # Normalize inputs
        inputsNorm = np.array([
            [(self.mins[i] if inputs[j, i] < self.mins[i] else self.maxes[i] if inputs[j, i] > self.maxes[i]
              else inputs[j, i] - self.means[i]) / self.ranges[i]
             for i in range(self.inputSize)]
            for j in range(inputs.shape[0])
        ])

        # Forward pass through the network
        for layer in range(self.numLayers - 1):
            inputsNorm = np.maximum(np.dot(self.weights[layer], inputsNorm.T) + self.biases[layer].reshape(-1, 1), 0)
        outputs = np.dot(self.weights[-1], inputsNorm) + self.biases[-1].reshape(-1, 1)

        # Undo output normalization
        outputs = outputs * self.ranges[-1] + self.means[-1]
        return outputs.T

    def num_inputs(self):
        """
        Get network input size.
        """
        return self.inputSize

    def num_outputs(self):
        """
        Get network output size.
        """
        return self.outputSize

File: julia/nnet.jl
module SimpleNNet

export NNet, evaluate_network, evaluate_network_multiple, num_inputs, num_outputs

# Define a simple structure to represent the neural network
mutable struct NNet
    weights::Array{Array{Float64, 2}}
    biases::Array{Array{Float64, 1}}
    numLayers::Int
    layerSizes::Array{Int}
    inputSize::Int
    outputSize::Int
    mins::Array{Float64}
    maxes::Array{Float64}
    means::Array{Float64}
    ranges::Array{Float64}

    # Constructor for loading the network from a file
    function NNet(file::String)
        # Open the file and skip comment lines
        f = open(file, "r")
        while startswith(peekline(f), "//")
            readline(f)
        end

        # Read basic network information
        metadata = split(readline(f), ",")
        numLayers = parse(Int, metadata[1])
        inputSize = parse(Int, metadata[2])
        outputSize = parse(Int, metadata[3])

        # Read layer sizes
        layerSizes = parse.(Int, split(readline(f), ","))

        # Skip a line
        readline(f)

        # Read input mins, maxes, means, and ranges
        mins = parse.(Float64, split(readline(f), ","))
        maxes = parse.(Float64, split(readline(f), ","))
        means = parse.(Float64, split(readline(f), ","))
        ranges = parse.(Float64, split(readline(f), ","))

        # Initialize weights and biases as empty arrays
        weights = []
        biases = []

        # Read weights and biases for each layer
        for layer in 1:numLayers
            weight_matrix = Array{Float64}(undef, layerSizes[layer+1], layerSizes[layer])
            for i in 1:layerSizes[layer+1]
                weight_matrix[i, :] = parse.(Float64, split(readline(f), ","))
            end
            push!(weights, weight_matrix)

            bias_vector = parse.(Float64, split(readline(f), ","))
            push!(biases, bias_vector)
        end

        close(f)

        return new(weights, biases, numLayers, layerSizes, inputSize, outputSize, mins, maxes, means, ranges)
    end
end

# Function to evaluate the network for one set of inputs
function evaluate_network(nnet::NNet, input::Array{Float64, 1})
    inputs = Array{Float64}(undef, nnet.inputSize)
    
    # Normalize inputs
    for i in 1:nnet.inputSize
        if input[i] < nnet.mins[i]
            inputs[i] = (nnet.mins[i] - nnet.means[i]) / nnet.ranges[i]
        elseif input[i] > nnet.maxes[i]
            inputs[i] = (nnet.maxes[i] - nnet.means[i]) / nnet.ranges[i]
        else
            inputs[i] = (input[i] - nnet.means[i]) / nnet.ranges[i]
        end
    end

    # Perform forward pass through the network
    for layer in 1:nnet.numLayers-1
        inputs = max.(nnet.weights[layer] * inputs .+ nnet.biases[layer], 0)
    end
    outputs = nnet.weights[end] * inputs .+ nnet.biases[end]

    # Undo output normalization
    for i in 1:nnet.outputSize
        outputs[i] = outputs[i] * nnet.ranges[end] + nnet.means[end]
    end

    return outputs
end

# Function to evaluate the network for multiple sets of inputs
function evaluate_network_multiple(nnet::NNet, input::Array{Float64, 2})
    numInputs = size(input, 2)
    inputs = Array{Float64}(undef, nnet.inputSize, numInputs)

    # Normalize inputs
    for i in 1:nnet.inputSize
        for j in 1:numInputs
            if input[i, j] < nnet.mins[i]
                inputs[i, j] = (nnet.mins[i] - nnet.means[i]) / nnet.ranges[i]
            elseif input[i, j] > nnet.maxes[i]
                inputs[i, j] = (nnet.maxes[i] - nnet.means[i]) / nnet.ranges[i]
            else
                inputs[i, j] = (input[i, j] - nnet.means[i]) / nnet.ranges[i]
            end
        end
    end

    # Perform forward pass through the network
    for layer in 1:nnet.numLayers-1
        inputs = max.(nnet.weights[layer] * inputs .+ nnet.biases[layer], 0)
    end
    outputs = nnet.weights[end] * inputs .+ nnet.biases[end]

    # Undo output normalization
    for i in 1:nnet.outputSize, j in 1:numInputs
        outputs[i, j] = outputs[i, j] * nnet.ranges[end] + nnet.means[end]
    end

    return outputs
end

# Get the number of inputs in the network
function num_inputs(nnet::NNet)
    return nnet.inputSize
end

# Get the number of outputs in the network
function num_outputs(nnet::NNet)
    return nnet.outputSize
end

end # module
